components/server/src/workspace/workspace-starter.ts

```diff
@@ -332,46 +338,123 @@ export class WorkspaceStarter {
-                log.error("actuallyStartWorkspace", err),
-            );
-            return { instanceID: instance.id };
-        }
+    public reconcileWorkspaceStart(_ctx: TraceContext, instanceId: string) {
+        const ctx = TraceContext.childContext("reconcileWorkspaceStart", _ctx);
+
+        // We try to acquire a mutex here, which we intend to hold until the workspace start request is sent to ws-manager.
+        // In case this container dies for whatever reason, the mutex is eventually released, and the instance can be picked up
+        // by another server process (cmp. controlAllOurStartingWorkspaces() above).
+        this.redisMutex
+            .using(
+                ["workspace-instance-start-" + instanceId],
+                5000, // after 5s without extension the lock is released
+                async (abortSignal) => {
+                    const instance = await this.workspaceDb.trace({}).findInstanceById(instanceId);
+                    if (!instance) {
+                        ctx.span.finish();
+                        throw new Error("cannot find workspace for instance");
+                    }
+                    const workspace = await this.workspaceDb.trace(ctx).findById(instance.workspaceId);
+                    if (!workspace) {
+                        throw new Error("cannot find workspace for instance");
+                    }
+                    const user = await this.userDB.trace(ctx).findUserById(workspace.ownerId);
+                    if (!user) {
+                        throw new Error("cannot find owner for workspace");
+                    }
+
+                    const envVars = await this.envVarService.resolveEnvVariables(
+                        user.id,
+                        workspace.projectId,
+                        workspace.type,
+                        workspace.context,
+                    );
+
+                    const forceRebuild = !!workspace.context.forceImageBuild;
+                    try {
+                        // if we need to build the workspace image we must not wait for actuallyStartWorkspace to return as that would block the
+                        // frontend until the image is built.
+                        const additionalAuth = await this.getAdditionalImageAuth(envVars);
+                        const needsImageBuild =
+                            forceRebuild ||
+                            (await this.needsImageBuild(ctx, user, workspace, instance, additionalAuth));
+                        if (needsImageBuild) {
+                            instance.status.conditions = {
+                                neededImageBuild: true,
+                            };
+                        }
+                        ctx.span?.setTag("needsImageBuild", needsImageBuild);
+                    } catch (err) {
+                        // if we fail to check if the workspace needs an image build (e.g. becuase the image builder is unavailable),
```

Comment:
Should we check abortSignal in such cases and simply exit? The same for exception handling  
in actuallyStartWorkspace? Otherwise we will make workspaces as stopped because of issues with Redis?

Reply:
I have to think about the error case a little bit more, was undecided yesterday.

Reply:
As we should prevent all DB updates once we no longer have the lock, I guarded the error handling as well. :+1: